### ISSUE TITLE
Support implicit `__class__` cells across nested and annotation scopes

### ISSUE THREAD

**Issue by charliermarsh** [astral-sh/ty#3850]
### Summary

In [astral-sh/ruff#26329](https://github.com/astral-sh/ruff/pull/26329), we added basic support for the implicit `__class__` closure cell, fixing [#3122](https://github.com/astral-sh/ty/issues/3122). We intentionally limited that support to direct method bodies and lazy scopes defined directly in a class body.

Here are some cases that remain unsupported (but are lower priority)...

Nested functions should inherit the cell from the enclosing method:

```py
class C:
    def method(self) -> None:
        def nested() -> None:
            reveal_type(__class__)  # expected: <class 'C'>
```

Type-alias annotation scopes should have access to the cell:

```py
class C:
    type Alias = __class__
    type GenericAlias[T] = __class__
```

Generic method bounds should have access to it:

```py
class C:
    def method[T: __class__](self) -> None: ...
```

Under Python 3.14's deferred annotation semantics, method annotations should also have access:

```py
class C:
    def method(
        self,
        value: __class__,
    ) -> __class__:
        ...
```

**Comment by carljm** [comment]
I think there's a notable priority differentiation among the above cases. The first case (nested functions) is supported by all other type checkers. None of the other listed cases are supported by any other type checker.
